pkg/restore/restore.go

```diff
@@ -458,92 +454,51 @@ func (ctx *restoreContext) execute() (Result, Result) {
 	// totalItems: previously discovered items, i: iteration counter.
 	totalItems, i, existingNamespaces := 0, 0, sets.NewString()
 
-	for _, selectedResource := range selectedResourceCollection {
+	// First restore only CRDs
+	crdResourceCollection, processedResources, w, e := ctx.getOrderedResourceCollection(
+		backupResources,
+		make([]restoreableResource, 0),
+		sets.NewString(),
+		[]string{"customresourcedefinitions"},
+		false,
+	)
+	warnings.Merge(&w)
+	errs.Merge(&e)
+
+	for _, selectedResource := range crdResourceCollection {
 		totalItems += selectedResource.totalItems
 	}
 
-	for _, selectedResource := range selectedResourceCollection {
-		groupResource := schema.ParseGroupResource(selectedResource.resource)
-
-		for namespace, selectedItems := range selectedResource.selectedItemsByNamespace {
-			for _, selectedItem := range selectedItems {
-				// If we don't know whether this namespace exists yet, attempt to create
-				// it in order to ensure it exists. Try to get it from the backup tarball
-				// (in order to get any backed-up metadata), but if we don't find it there,
-				// create a blank one.
-				if namespace != "" && !existingNamespaces.Has(selectedItem.targetNamespace) {
-					logger := ctx.log.WithField("namespace", namespace)
-
-					ns := getNamespace(
-						logger,
-						archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace),
-						selectedItem.targetNamespace,
-					)
-					_, nsCreated, err := kube.EnsureNamespaceExistsAndIsReady(
-						ns,
-						ctx.namespaceClient,
-						ctx.resourceTerminatingTimeout,
-					)
-					if err != nil {
-						errs.AddVeleroError(err)
-						continue
-					}
-
-					// Add the newly created namespace to the list of restored items.
-					if nsCreated {
-						itemKey := velero.ResourceIdentifier{
-							GroupResource: kuberesource.Namespaces,
-							Namespace:     ns.Namespace,
-							Name:          ns.Name,
-						}
-						ctx.restoredItems[itemKey] = struct{}{}
-					}
-
-					// Keep track of namespaces that we know exist so we don't
-					// have to try to create them multiple times.
-					existingNamespaces.Insert(selectedItem.targetNamespace)
-				}
+	for _, selectedResource := range crdResourceCollection {
+		var w, e Result
+		i, w, e = ctx.processSelectedResource(selectedResource, totalItems, i, existingNamespaces, update)
```

Comment:
I'd like to rename `i` to something more descriptive as this is a little confusing to follow - would something like `totalRestoredItems` work? `totalItems` could also be renamed to something like `totalItemsToRestore` to help distinguish what these two variables are tracking. I think this would also help with where these values are used in `processSelectedResource`.

As a follow up, rather than passing in `i` and making it the responsibility of `processSelectedResource` to update the value, could that function instead just return its own count of processed resources/restored items and then that could be added to the running tally in this outer scope (currently represented by `i`)? e.g.

```
restoredItems, w, e := ctx.processSelectedResource(selectedResource, totalItems, existingNamespaces, update)
totalRestoredItems += restoredItems
```

Reply:
So in digging into this further, `i` is actually not the number of restored items. It's an iterator count of processed items. If the item is already restored, then i increments while the number of items actually restored doesn't increase. Meanwhile, the actual list of restored items can increase outside of the loop iterations (namespaces added, plugin actions, etc.). I'm going to rename `i` to `processedItems` to make this clear.